docker-compose-rule-junit4/src/main/java/com/palantir/docker/compose/DockerComposeRule.java

```diff
@@ -1,160 +1,42 @@
 import org.junit.runners.model.Statement;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
 @Value.Immutable
 @CustomImmutablesStyle
-public abstract class DockerComposeRule extends ExternalResource {
-    private static final Logger log = LoggerFactory.getLogger(DockerComposeRule.class);
-
-    public static final Duration DEFAULT_TIMEOUT = Duration.standardMinutes(2);
-    public static final int DEFAULT_RETRY_ATTEMPTS = 2;
-
-    public DockerPort hostNetworkedPort(int port) {
-        return new DockerPort(machine().getIp(), port, port);
-    }
-
-    public abstract DockerComposeFiles files();
-
-    protected abstract List<ClusterWait> clusterWaits();
-
-    protected abstract List<EventConsumer> eventConsumers();
-
-    @Value.Default
-    public DockerMachine machine() {
-        return DockerMachine.localMachine().build();
-    }
-
-    @Value.Default
-    public ProjectName projectName() {
-        return ProjectName.random();
-    }
-
-    @Value.Default
-    public DockerComposeExecutable dockerComposeExecutable() {
-        return DockerComposeExecutable.builder()
-            .dockerComposeFiles(files())
-            .dockerConfiguration(machine())
-            .projectName(projectName())
-            .build();
-    }
-
-    @Value.Default
-    public DockerExecutable dockerExecutable() {
-        return DockerExecutable.builder()
-                .dockerConfiguration(machine())
-                .build();
-    }
-
-    @Value.Default
-    public Docker docker() {
-        return new Docker(dockerExecutable());
-    }
-
-    @Value.Default
-    public ShutdownStrategy shutdownStrategy() {
-        return ShutdownStrategy.KILL_DOWN;
-    }
-
-    @Value.Default
-    public DockerCompose dockerCompose() {
-        DockerCompose dockerCompose = new DefaultDockerCompose(dockerComposeExecutable(), machine());
-        return new RetryingDockerCompose(retryAttempts(), dockerCompose);
-    }
-
-    @Value.Default
-    public Cluster containers() {
-        return ImmutableCluster.builder()
-                .ip(machine().getIp())
-                .containerCache(new ContainerCache(docker(), dockerCompose()))
-                .build();
-    }
-
-    @Value.Default
-    protected int retryAttempts() {
-        return DEFAULT_RETRY_ATTEMPTS;
-    }
-
-    @Value.Default
-    protected boolean removeConflictingContainersOnStartup() {
-        return true;
-    }
-
-    @Value.Default
-    protected boolean pullOnStartup() {
-        return false;
-    }
-
-    @Value.Default
-    protected ReadableDuration nativeServiceHealthCheckTimeout() {
-        return DEFAULT_TIMEOUT;
-    }
-
-    @Value.Default
-    protected LogCollector logCollector() {
-        return new DoNothingLogCollector();
-    }
-
-    @Value.Derived
-    protected EventEmitter emitEventsFor() {
-        return new EventEmitter(eventConsumers());
-    }
-
+public abstract class DockerComposeRule extends DockerComposeManager implements TestRule {
```

Comment:
It's kind of sad we can't keep compat with `ExternalResource` here for those who have used the vintage runner. The reason we can't do it is because of inheritance - can we use delegation here instead? A benefit is that we could make the junit4/junit5 versions exposed as interfaces instead of classes, which would mean we can avoid using Javassist to make `EnhancedDockerComposeRule` as well. Also, subclassing immutables classes has always seemed a little weird to me. It also means we could make DockerComposeManager package private and reduce our API surface?

Reply:
Caught up with @ferozco over zoom, think we can make DockerComposeMananger private-ish anyway by moving it to another project which is not part of the api exposed by `docker-compose-rule-(junit4|jupiter)`.
WRT people who have migrated to using this via the vintage runner, there shouldn't be many cases of this, and we are planning to cut a major version anyway to make it super clear that things are a bit different.

It would be nice-ish to make these interfaces, but that also means we'd have to write a bunch of delegation code twice which is also not pretty :(

Reply:
I looked into moving parts of `docker-compose-rule-core` into a separate API package and making `docker-compose-rule-junit4` take an implementation dependency on `docker-compose-rule-core` and an api dependency on `docker-compose-rule-api`. It doesn't seem worth while since DCM re-exposes most of `docker-compose-rule-core`